gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/DatasetTest.java

```diff
@@ -0,0 +1,226 @@
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Iterator;
+
+public class DatasetTest {
+
+  private static final DatasetId DATASET_ID = DatasetId.of("dataset");
+  private static final DatasetInfo DATASET_INFO = DatasetInfo.builder(DATASET_ID).build();
+  private static final Field FIELD = Field.of("FieldName", Field.Type.integer());
+  private static final Iterable<BaseTableInfo> TABLE_INFO_RESULTS = ImmutableList.of(
+      TableInfo.builder(TableId.of("dataset", "table1"), Schema.of(FIELD)).build(),
+      ViewInfo.builder(TableId.of("dataset", "table2"), "QUERY").build(),
+      ExternalTableInfo.builder(TableId.of("dataset", "table2"),
+          ExternalDataConfiguration.of(ImmutableList.of("URI"), Schema.of(), FormatOptions.csv()))
+          .build());
+
+  private BigQuery bigquery;
+  private Dataset dataset;
+
+  @Before
+  public void setUp() throws Exception {
+    bigquery = createStrictMock(BigQuery.class);
+    dataset = new Dataset(bigquery, DATASET_INFO);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    verify(bigquery);
+  }
+
+  @Test
+  public void testInfo() throws Exception {
+    assertEquals(DATASET_INFO, dataset.info());
+    replay(bigquery);
+  }
+
+  @Test
+  public void testExists_True() throws Exception {
+    BigQuery.DatasetOption[] expectedOptions = {BigQuery.DatasetOption.fields()};
+    expect(bigquery.getDataset(DATASET_ID, expectedOptions)).andReturn(DATASET_INFO);
+    replay(bigquery);
+    assertTrue(dataset.exists());
+  }
+
+  @Test
+  public void testExists_False() throws Exception {
+    BigQuery.DatasetOption[] expectedOptions = {BigQuery.DatasetOption.fields()};
+    expect(bigquery.getDataset(DATASET_ID, expectedOptions)).andReturn(null);
+    replay(bigquery);
+    assertFalse(dataset.exists());
+  }
+
+  @Test
+  public void testReload() throws Exception {
+    DatasetInfo updatedInfo = DATASET_INFO.toBuilder().description("Description").build();
+    expect(bigquery.getDataset(DATASET_ID.dataset())).andReturn(updatedInfo);
+    replay(bigquery);
+    Dataset updatedDataset = dataset.reload();
+    assertSame(bigquery, updatedDataset.bigquery());
+    assertEquals(updatedInfo, updatedDataset.info());
+  }
+
+  @Test
+  public void testReloadNull() throws Exception {
+    expect(bigquery.getDataset(DATASET_ID.dataset())).andReturn(null);
+    replay(bigquery);
+    assertNull(dataset.reload());
+  }
+
+  @Test
+  public void testReloadWithOptions() throws Exception {
+    DatasetInfo updatedInfo = DATASET_INFO.toBuilder().description("Description").build();
+    expect(bigquery.getDataset(DATASET_ID.dataset(), BigQuery.DatasetOption.fields()))
+        .andReturn(updatedInfo);
+    replay(bigquery);
+    Dataset updatedDataset = dataset.reload(BigQuery.DatasetOption.fields());
+    assertSame(bigquery, updatedDataset.bigquery());
+    assertEquals(updatedInfo, updatedDataset.info());
+  }
+
+  @Test
+  public void testUpdate() throws Exception {
```

Comment:
Should we add a test where we try to call `update` with a `DatasetInfo` that has a different ID?